draft-ietf-tls-tls13.md

```diff
@@ -4696,6 +4641,196 @@ a context in all exporter computations, though the value could be empty.
+then a ClientHello may be received that is outside the window in the
+future, in which case it might be accepted for 1-RTT, causing a client retry,
+and then acceptable later for 0-RTT. This is another variant of
+the second form of attack described above.
+
+
+
+## Freshness Checks
+
+Because the ClientHello indicates the time at which the client sent
+it, it is possible to efficiently determine whether a ClientHello was
+likely sent reasonably recently and only accept 0-RTT for such a
+ClientHello, otherwise falling back to a 1-RTT handshake.
+This is necessary for the ClientHello storage mechanism
+described in {{client-hello-recording}} because otherwise the server
+needs to store an unlimited number of ClientHellos and is a useful optimization for
+single-use tickets because it allows efficient rejection of ClientHellos
+which cannot be used for 0-RTT.
+
+In order to implement this mechanism, a server needs to store the time
+that the server generated the session ticket, offset by an estimate of
+the round trip time between client and server. I.e.,
+
+~~~~
+    adjusted_creation_time = creation_time + estimated_RTT
+~~~~
+
+This value can be encoded in the ticket, thus avoiding the need to
+keep state for each outstanding ticket. The server can determine the
+client's view of the age of the ticket by subtracting the ticket's
+"ticket_age_add value" from the "obfuscated_ticket_age" parameter in
+the client's "pre_shared_key" extension. The server can determine the
+"expected arrival time" of the ClientHello as:
+
+~~~~
+    expected_arrival_time = adjusted_creation_time + clients_ticket_age
+~~~~
+
+When a new ClientHello is received, the expected_arrival_time is then
+compared against the current server wall clock time and if they differ
+by more than a certain amount, 0-RTT is rejected, though the 1-RTT
+handshake can be allowed to complete.
+
+There are several potential sources of error that might cause
+mismatches between the expected arrival time and the measured
+time. Variations in client and server clock
+rates are likely to be minimal, though potentially with gross time
+corrections.  Network propagation delays are the most likely causes of
+a mismatch in legitimate values for elapsed time.  Both the
+NewSessionTicket and ClientHello messages might be retransmitted and
+therefore delayed, which might be hidden by TCP. For clients
+on the Internet, this implies windows
+on the order of ten seconds to account for errors in clocks and
+variations in measurements; other deployment scenarios
+may have different needs. Clock skew distributions are not
+symmetric, so the optimal tradeoff may involve an asymmetric range
+of permissible mismatch values.
+
+Note that freshness checking alone is not sufficient to prevent replays
+because it does not detect them during the error window, which,
+depending on bandwidth and system capacity could include
+billions of replays in real-world settings.
```

Comment:
I would also add something along the lines of:
"Also note that this freshness checking is only done at the time the ClientHello is received, and not when later early application data records are received. After early data is accepted, records may continue to be streamed to the server over a longer time period."
I believe there was some previous discussion on the list about using the time window to determine when it is safe to resend rejected 0-RTT data, which is not actually sufficient by itself.